### ISSUE TITLE
Maybe incorrect inception layer, or I understand incorrectly

### ISSUE THREAD

**Issue by myme5261314** [cmusatyalab/openface#142]
@bamos @melgor , I'm still checking the `openface` code line by line, even dive into some relevant library it used. And I have a confusion about current Inception layer in the nn4.def.lua model for now.

  I don't know which one is the reason for the confusion: the issue of [doc](https://github.com/torch/nn/blob/master/doc/convolution.md#nn.SpatialMaxPooling) or [code](https://github.com/torch/nn/blob/master/SpatialMaxPooling.lua) of `nn.SpatialMaxPooling()`, the issue in [code](https://github.com/Element-Research/dpnn/blob/master/Inception.lua) or [doc](https://github.com/Element-Research/dpnn#nn.Inception) of `nn.Inception()`, just leave these links as reference.

Let me describe the inconsistency.

Here's part of GoogLeNet structure from `GoogLeNet paper`
> ![image](https://cloud.githubusercontent.com/assets/1814831/15530549/bb3b83c6-2287-11e6-9292-ba3eefedcea7.png)

Here's part of GoogLeNet (**3a, 3b, max-pool, 4a**) from the paper `GoogLeNet paper`.
> ![image](https://cloud.githubusercontent.com/assets/1814831/15530491/3ad488a4-2287-11e6-8a66-79023c1f9f1f.png)

So, the `maxpool` with stride size **2** is kind of component independent from Inception layer group right? Even though you could add this extra **stride 2 maxpool** layer as an optimal option, but I can't find corresponding code or doc for this option (refer to those links of `nn.Inception`).

So, I assume that implement the GoogLeNet style Inception layer might need explicit maxpool layer with **stride 2**. These maxpool layers is also needed for FaceNet NN2 and NN4 structure.

Here's part of NN2 structure from `FaceNet paper`.
> ![image](https://cloud.githubusercontent.com/assets/1814831/15530391/a21d86c4-2286-11e6-9904-5416223f81fc.png)

But let's check what the code is about 3a~4a for example.
```lua
   -- Inception (3a)
   net:add(nn.Inception{
     inputSize = 192,
     kernelSize = {3, 5},
     kernelStride = {1, 1},
     outputSize = {128, 32},
     reduceSize = {96, 16, 32, 64},
     pool = nn.SpatialMaxPooling(3, 3, 2, 2),
     batchNorm = true
   })

   -- Inception (3b)
   net:add(nn.Inception{
     inputSize = 256,
     kernelSize = {3, 5},
     kernelStride = {1, 1},
     outputSize = {128, 64},
     reduceSize = {96, 32, 64, 64},
     pool = nn.SpatialLPPooling(256, 2, 3, 3),
     batchNorm = true
   })

   -- Inception (3c)
   net:add(nn.Inception{
     inputSize = 320,
     kernelSize = {3, 5},
     kernelStride = {2, 2},
     outputSize = {256, 64},
     reduceSize = {128, 32, nil, nil},
     pool = nn.SpatialMaxPooling(3, 3, 2, 2),
     batchNorm = true
   })

   -- Inception (4a)
   net:add(nn.Inception{
     inputSize = 640,
     kernelSize = {3, 5},
     kernelStride = {1, 1},
     outputSize = {192, 64},
     reduceSize = {96, 32, 128, 256},
     pool = nn.SpatialLPPooling(640, 2, 3, 3),
     batchNorm = true
   })
```

The L2-pool's stride size seems to be 1 as default. The max-pool's stride size seems to be **2** as specified by code(refer to the doc and code of `nn.SpatialMaxPooling()`). (by the way, the doc of nn.Inception mention no argument named `pool`).

I'm really confused about what is actually running? the maxpool or l2-pool inside the Inception group should have stride size as **1**, right? Otherwise, there would be error running the code. With stride 2, the output of max-pool is 1/2 of input height and width, and its size will not fit into the concat.

And I tried modified the stride of max-pool to 1 or 3, the code all runs successfully, no error.
```lua
pool = nn.SpatialMaxPooling(3, 3, 1, 1), or
pool = nn.SpatialMaxPooling(3, 3, 3, 3),
```
It seems the pool argument have no effects, maybe it's entirely useless according to the code of `nn.Inception`? I'm not an expert of lua and torch. I can't confirm this.

So, the question is, where's the missing separate max-pool layer with stride size 2?
Please correct me if I'm somewhat wrong.

**Comment by myme5261314** [comment]
Never mind, I'm wrong. Since while the stride of `maxpool` is 2, the stride of `conv` is also 2.
This is equivalent to use both stride 1 and concat and then use `maxpool` with stride 2.
And this seems to consume less computation.

Close the issue.

**Comment by myme5261314** [comment]
Maybe I'm right for only one thing?
The `pool=nn.SpatialMaxPooling` is weird. The max-pooling for **3a** and **3c** should not be the same, right? So, I'm really confused what happened.

**Comment by melgor** [comment]
I'm rather confused about that topic too, but I think both (3a and 3c) can be the same.
But for sure, output dimension of "N2" model is the same like in FaceNet paper. 

"N4" have smaller output because of removing some extra layers in last Inception module. 

I believe that it would not have as much impact on final Accuracy, if there is any difference between paper and OpenFace.

**Comment by myme5261314** [comment]
@melgor , I'm not talking about accuracy since the experiment results have proven themselves.

Am I not explain it properly? I mean I can't understand the conflicts in 3a layer of NN4.

Imagine the input to the 3a layer is with size x\*\x*\y.
```lua
   -- Inception (3a)
   net:add(nn.Inception{
     inputSize = 192,
     kernelSize = {3, 5},
     kernelStride = {1, 1},
     outputSize = {128, 32},
     reduceSize = {96, 16, 32, 64},
     pool = nn.SpatialMaxPooling(3, 3, 2, 2),
     batchNorm = true
   })
```
what's the output of 
1. 1x1 conv -> 3x3 conv, stride **1** -> x\*x\*??
2. 1x1 conv -> 5x5 conv, stride **1** -> x\*x\*??
3. 1x1 conv -> x\*x\*\??
4. 3x3 maxpool, stride **2** -> 1x1 conv -> x/2\*x/2\*??

So how does the output of maxpool in 3a layer to `depth concat` with other outputs? Their size is incompatible.

While the kernelstride in 3c layer changes to {2, 2} which is compatible with other outputs.

**Comment by melgor** [comment]
I made a quick test.
First of all, it looks like Inception in dpnn have only 3 branches, like presented below:
```
nn.DepthConcat {
  input
    |`-> (1): nn.Sequential {
    |      [input -> (1) -> (2) -> (3) -> (4) -> (5) -> (6) -> output]
    |      (1): nn.SpatialConvolution(192 -> 96, 1x1)
    |      (2): nn.SpatialBatchNormalization
    |      (3): nn.ReLU
    |      (4): nn.SpatialConvolution(96 -> 128, 3x3, 1,1, 1,1)
    |      (5): nn.SpatialBatchNormalization                                                                                                                                                                                                                                  
    |      (6): nn.ReLU                                                                                                                                                                                                                                                       
    |    }                                                                                                                                                                                                                                                                    
    |`-> (2): nn.Sequential {                                                                                                                                                                                                                                                 
    |      [input -> (1) -> (2) -> (3) -> (4) -> (5) -> (6) -> output]                                                                                                                                                                                                        
    |      (1): nn.SpatialConvolution(192 -> 16, 1x1)                                                                                                                                                                                                                         
    |      (2): nn.SpatialBatchNormalization                                                                                                                                                                                                                                  
    |      (3): nn.ReLU                                                                                                                                                                                                                                                       
    |      (4): nn.SpatialConvolution(16 -> 32, 5x5, 1,1, 2,2)                                                                                                                                                                                                                
    |      (5): nn.SpatialBatchNormalization                                                                                                                                                                                                                                  
    |      (6): nn.ReLU                                                                                                                                                                                                                                                       
    |    }
    |`-> (3): nn.Sequential {
           [input -> (1) -> output]
           (1): nn.SpatialMaxPooling(3x3, 2,2)
         }
     ... -> output
}
```

The output from each branch is following (where input is 192x48x48 (just guess size))
1. 128x48x48
2. 32x48x48
3. 192x23x23
4. Merged:  352x48x48

So it looks like "3" have smaller dimension. In the DepthConcat implementation you will find a note explaining the behavior in such situation: [note](https://github.com/torch/nn/blob/master/DepthConcat.lua#L2). 

I must say that I do not get the implementation of Inception layer at 'dpnn' library. I think it should be different.

**Comment by myme5261314** [comment]
so this is different with the approach described in Inception paper. Now you helped me confirmed that "3" in 3a is with around 1/4 size. Is this really ok? Despite it might have no effects to result, I think using innovated version approach of Inception layer require explanation, right?

发自我的 iPhone

> 在 2016年5月31日，19:03，Bartosz Ludwiczuk <notifications@github.com> 写道：
> 
> I made a quick test.
> First of all, it looks like Inception in dpnn have only 3 branches, like presented below:
> 
> nn.DepthConcat {
>   input
>     |`-> (1): nn.Sequential {
>     |      [input -> (1) -> (2) -> (3) -> (4) -> (5) -> (6) -> output]
>     |      (1): nn.SpatialConvolution(192 -> 96, 1x1)
>     |      (2): nn.SpatialBatchNormalization
>     |      (3): nn.ReLU
>     |      (4): nn.SpatialConvolution(96 -> 128, 3x3, 1,1, 1,1)
>     |      (5): nn.SpatialBatchNormalization                                                                                                                                                                                                                                  
>     |      (6): nn.ReLU                                                                                                                                                                                                                                                       
>     |    }                                                                                                                                                                                                                                                                    
>     |`-> (2): nn.Sequential {                                                                                                                                                                                                                                                 
>     |      [input -> (1) -> (2) -> (3) -> (4) -> (5) -> (6) -> output]                                                                                                                                                                                                        
>     |      (1): nn.SpatialConvolution(192 -> 16, 1x1)                                                                                                                                                                                                                         
>     |      (2): nn.SpatialBatchNormalization                                                                                                                                                                                                                                  
>     |      (3): nn.ReLU                                                                                                                                                                                                                                                       
>     |      (4): nn.SpatialConvolution(16 -> 32, 5x5, 1,1, 2,2)                                                                                                                                                                                                                
>     |      (5): nn.SpatialBatchNormalization                                                                                                                                                                                                                                  
>     |      (6): nn.ReLU                                                                                                                                                                                                                                                       
>     |    }
>     |`-> (3): nn.Sequential {
>            [input -> (1) -> output]
>            (1): nn.SpatialMaxPooling(3x3, 2,2)
>          }
>      ... -> output
> }
> The output from each branch is following (where input is 192x48x48 (just guess size))
> 1. 128x48x48
> 2. 32x48x48
> 3. 192x23x23
> 4. Merged: 352x48x48
> 
> So it looks like "3" have smaller dimension. In the DepthConcat implementation you will find a note explaining the behavior in such situation: note.
> 
> I must say that I do not get the implementation of Inception layer at 'dpnn' library. I think it should be different.
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

**Comment by bamos** [comment]
Hi @myme5261314 - great find! I didn't intentionally make the sizes different and I assumed they were all the same size since `dpnn.Inception`/DepthConcat didn't output any warnings. We should try re-training with this fixed.

Hi @melgor - what part of `dpnn.Inception` do you think should be different? I think this problem is just from how I'm using it.

\cc @nicholas-leonard - I accidentally made the output sizes different in my use of `dpnn.Inception`. What do you think about adding a check/warning so this is more visible?

-Brandon.

**Comment by nicholas-leonard** [comment]
@bamos Yeah we could add a warning. I will be interested to see how this changes the open face results.

@melgor The origin idea of DepthConcat was to support concatenation of different height/width without requiring a costly resize. I think the original paper was actually quite vague on how they handle depth-concatenation.

**Comment by myme5261314** [comment]
Actually, there is no great improvement in result, since I had implement the correct approach in my 'facenet' repo via tensorflow and finished some experiments. You could still try again. By the way, I'm really not satisfied with current 93% around accuracy, @melgor, is there any progress on improving that? I mean somebody said he could achieve 97.5% by himself, right? Have you confirmed that? Seems he didn't plan to share his structure to strangers.

发自我的 iPhone

**Comment by melgor** [comment]
@bamos Sorry, my test have a small bug. Correct 3a architecture looks like that:
```
nn.DepthConcat {
  input
    |`-> (1): nn.Sequential {
    |      [input -> (1) -> (2) -> (3) -> (4) -> (5) -> (6) -> output]
    |      (1): nn.SpatialConvolution(192 -> 96, 1x1)
    |      (2): nn.SpatialBatchNormalization
    |      (3): nn.ReLU
    |      (4): nn.SpatialConvolution(96 -> 128, 3x3, 1,1, 1,1)
    |      (5): nn.SpatialBatchNormalization
    |      (6): nn.ReLU
    |    }
    |`-> (2): nn.Sequential {
    |      [input -> (1) -> (2) -> (3) -> (4) -> (5) -> (6) -> output]
    |      (1): nn.SpatialConvolution(192 -> 16, 1x1)
    |      (2): nn.SpatialBatchNormalization
    |      (3): nn.ReLU
    |      (4): nn.SpatialConvolution(16 -> 32, 5x5, 1,1, 2,2)
    |      (5): nn.SpatialBatchNormalization
    |      (6): nn.ReLU
    |    }
    |`-> (3): nn.Sequential {
    |      [input -> (1) -> (2) -> (3) -> (4) -> output]
    |      (1): nn.SpatialMaxPooling(3x3, 2,2)
    |      (2): nn.SpatialConvolution(192 -> 32, 1x1)
    |      (3): nn.SpatialBatchNormalization
    |      (4): nn.ReLU
    |    }
    |`-> (4): nn.Sequential {
           [input -> (1) -> (2) -> (3) -> output]
           (1): nn.SpatialConvolution(192 -> 64, 1x1)
           (2): nn.SpatialBatchNormalization
           (3): nn.ReLU
         }
     ... -> output
}
```
Outputs:
1. 128x48x48
2. 32x48x48
3. 32x23x23
4. 64x48x48
5. Merged: 256x48x48

And it is look ok (same like in paper). So there is no bug in model architectures.

@myme5261314  For the progess, I do not have enough time to make some experiments (and computation power).  But the thing which are worth trying are:
1. Learn nn2 model (224x224 input) at CASIA + VGG database
2. Learn nn4 model for classification, then use OpenFace for Triplets (the same methodology like guy with 97.5%)
3. Add to OpenFace "Join Classification and Verification" by adding ParallelCriteria with SoftMax and TripletEmbeddingCriterion (there were some papers which claim that it helped a lot compared to pure Triplets)

**Comment by myme5261314** [comment]
I don't understand why you are saying `same like in paper`, the
`width` and `height` in "3" is different with all others. I know in
this situation the code wouldn't give out error. But you should know
this is because the `depthconcat` have optimized this for you. No
error doesn't means it's the same with the paper.

If you think about more about it, you will find it's not the same. I mean,
> "stride 2 -> padding 0 outside -> stride 2" is not the same with `stride 1 -> stride 2"

There will be many padding 0 remains that have squeezed out position
that should have non-zero features!

For the 97.5%, have you re-implemented that result? I mean, in certain
issue, you have said you couldn't achieved that result, nor do @bamos.
Am I misunderstanding?

**Comment by melgor** [comment]
You are right, that this is not exactly the same architecture, but I thought that number of branches differ, but it is ok at 3a.

About Parameters:
MaxPooling:  should be ``` nn.SpatialMaxPooling(3, 3, 1, 1,1,1)``` , so stride 1 and pad 1.
LPPolling:     should be ```nn.SpatialLPPooling(960, 2, 3, 3,1,1) ``` , so stride 1. Torch implementation does not support padding.

With this changes, we have all right sizes with Inception layers containing  SpatialMaxPooling and slight smaller where we use LPPooling (because of no pad).

About 97.5%, I did not achieve such results, but I get ~95% where I just learn nn4 on a classification task. For sure, I could not achieve such results using Triplets. I did not try then using Triplets on 
pretrained model. But, as I said earlier, it is worth to try this 3 approaches which I mention before.

**Comment by bamos** [comment]
Also on improving accuracy, somebody recently posted an improved alignment technique to the mailing list that looks promising: https://groups.google.com/forum/#!topic/cmu-openface/aQuYBLcVKAo. I just started training a new model with it.

**Comment by myme5261314** [comment]
Interesting! Seems lack of CV experience makes me no aware of correct alignment appearances. I thought that appearance before was the normal result. = =!

I do think that will be a great improvement to the accuracy. I will be on a long holiday a few days later. The remaining days is not enough for me to go through all 3 approaches plus this approach. Please let us know the latest significant progress. I will try my best to run experiments for at most 2 approaches before I leave for vacation.

And about 1800 images per batch, it seems no significant effects on accuracy, and even converge very slowly. Or maybe it's the inappropriate alignment limited its effects. I think I should investigate more after alignment is correct.

@melgor and @bamos Thanks for all your information. These information will save me so much time.

Cheers. :-)

**Comment by bamos** [comment]
Hi @melgor and @myme5261314 - I added output size checks to dpnn and sent in a PR at https://github.com/Element-Research/dpnn/pull/63. I think I've corrected the Inception towers in OpenFace and have started training a new model. Can you all double-check my changes in https://github.com/cmusatyalab/openface/commit/49fffb36710086e0a0540d194fd53e7954fd563c and make sure I didn't overlook anything this time?

-Brandon.

**Comment by bamos** [comment]
I think my changes from 2 months ago were correct and I'm closing this issue. Let me know if you notice anything wrong in the updated definitions.

-Brandon.
